gm4/plugins/output.py

```diff
@@ -51,46 +59,126 @@ def release(ctx: Context):
+					"version_type": "release",
+					"loaders": ["datapack"],
+					"featured": False,
+					"project_id": modrinth_id,
+					"file_parts": [file_name],
+				}),
+				file_name: file_bytes,
+			})
+			if not (200 <= res.status_code < 300):
+				print(f"[GM4] [Modrinth] Failed to publish new version version: {res.status_code} {res.text}")
+				return
+			print(f"[GM4] [Modrinth] Successfully published {res.json()['name']}")
+
+def publish_smithed(ctx: Context, release_dir: Path, file_name: str):
+	"""Attempts to publish pack to smithed"""
+	smithed = ctx.meta.get("smithed", None)
+	auth_token = os.getenv(SMITHED_AUTH_KEY, None)
+	if smithed and auth_token and ctx.project_version:
+		version = version = ctx.cache["gm4_manifest"].json["modules"].get(ctx.project_id, {}).get("version", None)
+		smithed_id = smithed["pack_id"]
+
+		# get existing versions
+		res = requests.get(f"{SMITHED_API}/packs/{smithed_id}/versions")
+		if not (200 <= res.status_code < 300):
+			if res.status_code == 404:
+				print(f"[GM4] [Smithed] Cannot publish to smithed project {smithed_id} as it doesn't exist.")
+			else:
+				print(f"[GM4] [Smithed] Failed to get project versions: {res.status_code} {res.text}")
+			return
+		
 		project_data = res.json()
-		matching_version = next((v for v in project_data if v["version_number"] == ctx.project_version), None)
+		matching_version = next((v for v in project_data if v["name"] == version), None)
 		if matching_version is not None:
 			return
-
-		with open(release_dir / file_name, "rb") as f:
-			file_bytes = f.read()
-
-		game_versions = modrinth.get("minecraft", SUPPORTED_GAME_VERSIONS)
-		res = requests.post(f"{MODRINTH_API}/version", headers={'Authorization': auth_token}, files={
-			"data": json.dumps({
-				"name": f"{ctx.project_name} {ctx.project_version}",
-				"version_number": ctx.project_version,
-				"dependencies": [],
-				"game_versions": game_versions,
-				"version_type": "release",
-				"loaders": ["minecraft"],
-				"featured": False,
-				"project_id": modrinth_id,
-				"file_parts": [file_name],
-			}),
-			file_name: file_bytes,
-		})
+		
+		# remove other existing versions for that mc version
+		mc_version_matching_version = (v["name"] for v in project_data if v['supports'][0] == SUPPORTED_GAME_VERSIONS[0]) # NOTE smithed currently only supports one game version
+		for v in mc_version_matching_version:
+			res = requests.delete(f"{SMITHED_API}/packs/{smithed_id}/versions/{v}", params={'token': auth_token})
+			if not (200 <= res.status_code < 300):
+				print(f"[GM4] [Smithed] Failed to delete version version: {res.status_code} {res.text}")
+			print(f"[GM4] [Smithed] {res.text}")
+		
+		# post new version
+		res = requests.post(f"{SMITHED_API}/packs/{smithed_id}/versions",
+		    params={'token': auth_token, 'version': version},
+			json={"data":{
+				"downloads":{
+					"datapack": f"https://raw.githubusercontent.com/Gamemode4Dev/GM4_Datapacks/release/{SUPPORTED_GAME_VERSIONS[0]}/{file_name}",
```

Comment:
Use `os.getenv("VERSION", "1.19")` for this. It's possible our supported versions in the future could be `[1.19.3, 1.19.4]` or similar.

Reply:
Will do. That is much nicer

Smithed right now only has options for [1.18, 1.18.1, 1.18.2, 1.19] for its supported versions, hence why I've used this kinda ugly `SUPPORTED_GAME_VERSIONS[0]`. I'm hoping to let it just use the whole list in the future when smithed comes out of beta again.